File: src/main/java/com/sothawo/springdataelastictest/scriptandruntime/Person.java
package com.sothawo.springdataelastictest.scriptandruntime;

import org.springframework.data.annotation.Id;
import org.springframework.data.elasticsearch.annotations.DateFormat;
import org.springframework.data.elasticsearch.annotations.Document;
import org.springframework.data.elasticsearch.annotations.Field;
import org.springframework.data.elasticsearch.annotations.ScriptedField;
import org.springframework.lang.Nullable;

import java.time.LocalDate;
import java.time.format.DateTimeFormatter;

import static org.springframework.data.elasticsearch.annotations.FieldType.*;

import java.lang.Integer;

@Document(indexName = "sar-person")
public record Person(
        @Id
        @Nullable
        String id,
        @Field(type = Text)
        String lastName,
        @Field(type = Text)
        String firstName,
        @Field(type = Keyword) String gender,
        @Field(type = Date, format = DateFormat.basic_date)
        LocalDate birthDate,
        @Nullable
        @ScriptedField Integer age
) {
    public Person(String id,String lastName, String firstName, String gender, String birthDate) {
        this(id,
                lastName,
                firstName,
                gender,
                LocalDate.parse(birthDate, DateTimeFormatter.ISO_LOCAL_DATE),
                null);
    }
}

File: src/main/kotlin/com/sothawo/springdataelastictest/Person.kt
/*
 * (c) Copyright 2019 sothawo
 */
package com.sothawo.springdataelastictest

import org.springframework.data.annotation.Id
import org.springframework.data.elasticsearch.annotations.Document
import org.springframework.data.elasticsearch.annotations.Field
import org.springframework.data.elasticsearch.annotations.FieldType

/**
 * @author P.J. Meisch (pj.meisch@sothawo.com)
 */
@Document(indexName = "person")
class Person {
    @Id
    var id: Long? = null

    @Field(value = "last-name", type = FieldType.Text)
    var lastName: String? = null

    @Field(name = "first-name", type = FieldType.Text, fielddata = true)
    var firstName: String? = null

    @Field(type = FieldType.Nested)
    private val movies: List<Movie>? = null

    override fun toString(): String {
        return "Person{" +
                "id=" + id +
                ", lastName='" + lastName + '\''.toString() +
                ", firstName='" + firstName + '\''.toString() +
                '}'.toString()
    }
}
